### PR TITLE
[Windows] Projects without ".idea\.name" file wasn't auto-opened in the PhpStorm 2026.2

### PR DESCRIPTION
**Pull request by shaffe-fr**
PhpStorm 2026.2 broke the Windows protocol handler in two ways:

- `--line` is now ignored unless the project path is passed as an argument. The project is detected by walking up from the file, but it looked for the `.idea/.name` file which is no longer always present. It now looks for the `.idea` folder instead.
- `shell.Exec` does not launch the Toolbox `.cmd` wrapper properly, so PhpStorm never opened. Switched to `shell.Run`.

Tested on Windows 11 with Toolbox 3.6.2 and PhpStorm 2026.2.0.1.

### CODE REVIEW

**Comment by aik099** [comment]
I lots of changes that are loosely connected together and hardly reviewable.

Could you please create smaller on-topic PRs with their own reasoning and code changes, e.g.:

> change of `shell.Exec` to `shell.Run`

What is that for?

> improved project detection to support PhpStorm 2026.2+

The real fix (as I've guessed from your changes) was to improve project detecting by checking `.idea` folder (always present) instead of `.idea/.name` (present only when the project name in PhpStorm differs from the project folder name). That would solve the #68 issue as well.

> JetBrains Toolbox 2.0 support

I wonder how with all these path changes the PhpStorm downloaded at ToolBox 1.x path now event works in Toolbox 2.x paths?

**Comment by shaffe-fr** [comment]
I understand, I hesitated before committing all these changes. I'll create different PRs and trim extra from this one.

**Comment by aik099** [comment]
If you'll find any other topics that I haven't mentioned before that you could extract into separate PRs, then please do so.

The testing part would be hard. I haven't verified that by hand, but AI review reported BC breaks (see below).


## Claude Code generated review (double-check for issues 😄)

Reviewed the diff (`gh pr diff 69`) against this project's backward-compatibility principle: the `phpstorm://` URL format is the stable public API, but the underlying Toolbox/PhpStorm detection logic constantly changes upstream — so fixes here should add fallback branches, not replace/reorder existing ones, since users can't always upgrade OS/PhpStorm/Toolbox in lockstep with this repo.

Two real regressions, plus one lower-risk item:

**1. Favorite-channel selection silently breaks under Toolbox 2.0+ (real regression)**

PR #48 added `toolbox_update_channel_dir` specifically so users with two installed channels (e.g. PhpStorm stable + EAP) can pick which one `phpstorm://` launches. The new `configureToolboxSettings` state.json branch just does:
```js
for (var i = 0; i < tools.length; i++) {
    if (tools[i].toolId == 'PhpStorm') { ... return; }
}
```
— it takes whichever `tools[]` entry comes first and never looks at `settings.toolbox_update_channel_dir` or calls `getFavoritePhpStormChannel()`. Anyone who set that option and upgrades to Toolbox 2.0+ silently loses channel selection with no error, no fallback — just wrong-IDE-launches. `getFavoritePhpStormChannel()`'s new "Toolbox 2.0+ `channels/` dir" branch also appears to be dead code as written — nothing in the new primary path calls it.

**2. Toolbox shell-script preference is reordered, not preserved (regression against PR #58/#59)**

Original `getPhpStormCommandPath()` tried `getToolboxShellScript()` *first*, falling back to raw exe resolution — deliberately, per the existing comment: "the JetBrains Toolbox Shell Script is clever enough to autofocus PhpStorm window after opening a file in it." This PR flips the order: it now resolves the raw `.exe` path from `state.json` first, and only falls back to the shell script if that fails. For any Toolbox 2.0+ user who configured `toolbox_shell_script` (a feature added specifically to fix window-focus problems), this silently reverts to the old direct-launch behavior that #58/#59 fixed — no error, just the previously-fixed bug coming back for a subset of users.

Suggest this should try the shell script first exactly like before, and only use the new direct-`state.json` resolution as an additional fallback when no shell script is configured/found.

**3. Lower risk: `shell.Exec` → `shell.Run(command, 1, false)`**

Justified by the PR (Exec doesn't reliably launch GUI exes in newer Windows/PhpStorm), but window style `1` (normal) means launching via a `.cmd` toolbox shell script will flash a console window that `Exec` never showed. Cosmetic, not a functional break — worth a mention in review but not blocking.

**Not a concern:** `.idea\.name` → `.idea` folder-exists check for project detection is a strict superset (anything with `.name` also has the folder), so no regression there. This part looks like a durable, correct fix for #68's project-detection gap (`.idea\.name` only exists when the project name diverges from the folder name, so it under-detects by default), independent of the PhpStorm 2026.2 CLI bug that happened to surface it.

**Comment by shaffe-fr** [comment]
I understand, I hesitated before committing all these changes. I'll create different PRs and trim this one as well.
For this one, two things were required: the `.idea` detection improvement and now the `--line`argument must come first.

**Comment by aik099** [comment]
I've just merged #70 with my CLAUDE.md file. So if you happen to use Claude Code, that will help to steer it in the right direction.

**Comment by shaffe-fr** [comment]
Since PhpStorm 2026.2, the protocol handler no longer opens files on Windows. Two changes fix it.

First, `--line` is now ignored unless the project path is also passed as an argument. The script walks up the directories to find the project root, but it was looking for the `.idea/.name` file, which is not always present anymore. Looking for the `.idea` folder instead is more reliable.

Second, with Toolbox installed `getPhpStormCommandPath()` returns the generated `.cmd` launcher, and `shell.Exec` does not start a `.cmd` wrapper properly, so nothing happened. `shell.Run` handles it:

```js
shell.Run(command, 1, false);
```

Both changes stay backward compatible with standalone installs and older Toolbox versions.

Tested on Windows 11, Toolbox 3.6.2, PhpStorm 2026.2.0.1.

Regards

**Comment by aik099** [comment]
> but it was looking for the .idea/.name file, which is not always present anymore. 

It is never reliable because `.idea/.name` file is ONLY present when the folder name of a project differs from the project name in PhpStorm. I wonder why this hasn't failed earlier.

> and shell.Exec does not start a .cmd wrapper properly

How was it working before?

File: PhpStorm Protocol (Win)/run_editor.js [hunk 1]
```diff
@@ -62,7 +62,9 @@ if (match) {
         while (search_path.lastIndexOf('\\') !== -1) {
             search_path = search_path.substring(0, search_path.lastIndexOf('\\'));
 
-            if (file_system.FileExists(search_path + '\\.idea\\.name')) {
+            // The .idea/.name file is not always present, look for the .idea folder instead.
+            // PhpStorm 2026.2+ ignores --line unless the project path is passed as an argument.
```
Comment (aik099) [inline PhpStorm Protocol (Win)/run_editor.js]: Explanatory comment doesn't make sense, because it explains why old code (no longer visible in the file) wasn't working.

**Comment by shaffe-fr** [comment]
That's a good question, it wasn't working with Exec, I'll dig some more.

File: PhpStorm Protocol (Win)/run_editor.js [hunk 2]
```diff
@@ -62,7 +62,9 @@ if (match) {
...
+            if (file_system.FolderExists(search_path + '\\.idea')) {
                 project = search_path;
                 break;
             }
```
Comment (aik099) [inline PhpStorm Protocol (Win)/run_editor.js:68]: Maybe we need to show a message back to the user telling that opening a file outside a PhpStorm project doesn't work in PhpStorm 2026.2 in the `else` branch?

---

Opening file outside of project with `--line` specified would fail as well.

File: PhpStorm Protocol (Win)/run_editor.js [hunk 3]
```diff
@@ -84,7 +86,9 @@ if (match) {
         .replace(/%project%/g, project)
         .replace(/\//g, '\\');
 
-    shell.Exec(command);
+    // Use Run instead of Exec: Exec does not launch the Toolbox .cmd wrapper properly.
```
Comment (aik099) [inline PhpStorm Protocol (Win)/run_editor.js]: Why it worked before with `.Exec(...)`?

**Review by aik099 — CHANGES_REQUESTED** [review]

**Comment by aik099** [comment]
> That's a good question, it wasn't working with Exec, I'll dig some more.

I have [an enormous testing matrix](https://github.com/user-attachments/files/30371136/vm-test-matrix.md) that Claude Code built. I was planning to use for manual testing and identifying all the PhpStorm-vs-Toolbox inconsistencies. You can use that for testing as well. Although, I admit that I have Windows installed in VM that allows me to easily rollback any changes and start fresh with a new PhpStorm/Toolbox version for testing. That test matrix is based on that.

**Comment by shaffe-fr** [comment]
Ok, I pushed the minimal needed change so it works again for me.

Sorry for the back and forth.

**Comment by aik099** [comment]
So did the `shell.Exec` work for you at the end?

**Comment by aik099** [comment]
I've some tests locally on Windows 11 ARM with JetBrains Toolbox 3.x and PhpStorm 2026.1 and PhpStorm 2026.2 and they worked with the change you proposed. No huge toolbox change-related code (in you previously did) was needed.

How is that even possible? Am I missing some test scenario where it saved the day?

**Comment by shaffe-fr** [comment]
I guess the edge case is so small that I was  lost across all the testing I did

**Comment by aik099** [comment]
What do you think about improving this PR and showing an alert box for a case when for a given file path no IDE project (having `.idea` folder inside) was detected? This was exactly the case you are fixing, but if somebody tries to open a file outside a project, he'll get the same problem again with PHPStorm 2026.2 and won't be notified in any way that this flow isn't supported.

**Comment by aik099** [comment]
FYI I've made lots of changes, so I recommend rebasing your PR branch on top of the current master branch:

- #71 — removed the non-functional `window_title`/`AppActivate` window-focus code (confirmed unnecessary; PhpStorm's own single-instance relaunch handles focus regardless of launch path).
- #72 — fixed a ~2-year-old regression in Toolbox 1.x support, introduced by #50 when it added `state.json` support (the v1 folder-scan's result was being silently overwritten before ever reaching `shell.Exec`).
- #74 — fixed `toolbox_update_channel_dir` being ignored by the `state.json` resolution path on Toolbox 2.0+/3.x with multiple installed channels — another gap #50 never wired up.
- #31 — fixed Toolbox v1 picking an older PhpStorm build over a newer one when multiple versions are kept (a long-standing bug reported back in the original #30 thread); rebased and reimplemented since the 2019 diff no longer applied.
- #73 — updated `CLAUDE.md` with what we learned investigating all of the above.
- #75 — added a Compatibility section to the README covering tested Toolbox/PhpStorm version coverage.
- #76 — added `CONTRIBUTING.md` documenting how testing actually happens for Windows changes.

None of these touch the `.idea` folder detection change this PR is specifically about.

**Comment by aik099** [comment]
Actually, I'll merge this PR. Then you'll on new `master` branch if the code changes (e.g. toolbox v3 support) you've submitted earlier are still needed.

Merging. Thank you @shaffe-fr .
